kpi/models/asset.py

```diff
@@ -1000,24 +1028,138 @@ def clean_up_table():
                     _("Can not assign '{}' permission. "
                       "Partial permissions are missing.".format(perm)))
 
-            new_partial_perms = {}
+            new_partial_perms = defaultdict(list)
+            in_op = MongoHelper.IN_OPERATOR
+
             for partial_perm, filters in partial_perms.items():
-                implied_perms = [implied_perm
-                                 for implied_perm in
-                                 self.get_implied_perms(partial_perm)
-                                 if implied_perm.endswith(SUFFIX_SUBMISSIONS_PERMS)
-                                 ]
-                implied_perms.append(partial_perm)
+
+                if partial_perm not in new_partial_perms:
+                    new_partial_perms[partial_perm] = filters
+
+                implied_perms = [
+                    implied_perm
+                    for implied_perm in self.get_implied_perms(partial_perm)
+                    if implied_perm.endswith(SUFFIX_SUBMISSIONS_PERMS)
+                ]
+
                 for implied_perm in implied_perms:
-                    if implied_perm not in new_partial_perms:
-                        new_partial_perms[implied_perm] = []
-                    new_partial_perms[implied_perm] += filters
+
+                    if (
+                        implied_perm not in new_partial_perms
+                        and implied_perm in partial_perms
+                    ):
+                        new_partial_perms[implied_perm] = partial_perms[implied_perm]
+
+                    new_partial_perm = new_partial_perms[implied_perm]
+                    # Trivial case, i.e.: permissions are built with front end.
+                    # All permissions have only one filter and the same filter
+                    # Example:
+                    # ```
+                    # partial_perms = {
+                    #   'view_submissions' : [
+                    #       {'_submitted_by': {'$in': ['johndoe']}
+                    #   ],
+                    #   'delete_submissions':  [
+                    #       {'_submitted_by': {'$in': ['quidam']}
+                    #   ]
+                    # }
+                    # ```
+                    # should give
+                    # ```
+                    # new_partial_perms = {
+                    #   'view_submissions' : [
+                    #       {'_submitted_by': {'$in': ['johndoe', 'quidam']}
+                    #   ],
```

Comment:
answer: some important organizations like to assign 300+ users to a single project :upside_down_face: an this will make it easier to inspect running queries, e.g. when troubleshooting performance issues